res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js

```diff
@@ -0,0 +1,1336 @@
+      } else {
+        script.triggerControl(samplerGroup, "eject");
+      }
+    } else {
+      if (engine.getValue(samplerGroup, "track_loaded") === 0) {
+        script.triggerControl(samplerGroup, "LoadSelectedTrack");
+      } else {
+        script.triggerControl(samplerGroup, "cue_gotoandplay");
+      }
+    }
+  }
+}
+
+TraktorS2MK2.samplerModeButton = function(field) {
+  if (field.value === 0) {
+    return;
+  }
+  var padMode = TraktorS2MK2.current_pad_mode[field.group];
+  if (padMode !== TraktorS2MK2.pad_modes.sampler) {
+    TraktorS2MK2.setPadMode(field.group, TraktorS2MK2.pad_modes.sampler);
+    TraktorS2MK2.controller.setOutput(field.group, "!remix_button", 0x7f, false);
+    TraktorS2MK2.controller.setOutput(field.group, "!flux_button", 0x00, !TraktorS2MK2.freeze_lights);
+  } else {
+    TraktorS2MK2.setPadMode(field.group, TraktorS2MK2.pad_modes.hotcue);
+    TraktorS2MK2.controller.setOutput(field.group, "!remix_button", 0x00, !TraktorS2MK2.freeze_lights);
+  }
+}
+
+TraktorS2MK2.introOutroModeButton = function(field) {
+  if (field.value === 0) {
+    return;
+  }
+  var padMode = TraktorS2MK2.current_pad_mode[field.group];
+  if (padMode !== TraktorS2MK2.pad_modes.intro_outro) {
+    TraktorS2MK2.setPadMode(field.group, TraktorS2MK2.pad_modes.intro_outro);
+    TraktorS2MK2.controller.setOutput(field.group, "!flux_button", 0x7f, false);
+    TraktorS2MK2.controller.setOutput(field.group, "!remix_button", 0x00, !TraktorS2MK2.freeze_lights);
+  } else {
+    TraktorS2MK2.setPadMode(field.group, TraktorS2MK2.pad_modes.hotcue);
+    TraktorS2MK2.controller.setOutput(field.group, "!flux_button", 0x00, !TraktorS2MK2.freeze_lights);
+  }
+}
+
+TraktorS2MK2.loopIn = function(field) {
+  var group = field.id.split(".")[0];
+  engine.setValue(group, "loop_in", field.value);
+}
+
+TraktorS2MK2.loopOut = function(field) {
+  var group = field.id.split(".")[0];
+  engine.setValue(group, "loop_out", field.value);
+}
+
+TraktorS2MK2.samplerActivateButton = function(field) {
+  if (field.value > 0) {
+    TraktorS2Mk2.controller.samplers_active[field.group] = !TraktorS2Mk2.controller.samplers_active[field.group];
+  }
+}
+
+TraktorS2MK2.connectEffectButtonLEDs = function(effectUnitGroup) {
+  TraktorS2MK2.effectButtonLEDconnections[effectUnitGroup].forEach(function(connection) {
+    connection.disconnect();
+  });
+  
+  var focusedEffect = engine.getValue(effectUnitGroup, "focused_effect");
+  var makeButtonLEDcallback = function(effectNumber) {
+    return function(value, group, control) {
+      TraktorS2MK2.controller.setOutput(effectUnitGroup, "!effectbutton" + effectNumber,
+        value === 1 ? 0x7f : 0, !TraktorS2MK2.freeze_lights);
+    };
+  };
+  
+  // FIXME: Why do the LEDs flicker?
+  TraktorS2MK2.freeze_lights = true;
+  for (var i = 0; i < 3; i++) {
+    var effectGroup;
+    var key;
+    if (focusedEffect === 0) {
+      effectGroup = effectUnitGroup.slice(0, -1) + "_Effect" + (i+1) + "]";
```

Comment:
I thnk this is the string manipulation you mentioned before.  Instead, you could just do a simple regex instead (regexes are very fast) https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace and https://opnsrce.github.io/javascript-performance-tip-precompile-your-regular-expressions